core/src/main/java/org/elasticsearch/index/translog/Translog.java

```diff
@@ -1437,30 +1487,51 @@ public void rollGeneration() throws IOException {
      * Trims unreferenced translog generations by asking {@link TranslogDeletionPolicy} for the minimum
      * required generation
      */
-    public void trimUnreferencedReaders() {
+    public void trimUnreferencedReaders() throws IOException {
         try (ReleasableLock ignored = writeLock.acquire()) {
             if (closed.get()) {
                 // we're shutdown potentially on some tragic event, don't delete anything
                 return;
             }
             long minReferencedGen = deletionPolicy.minTranslogGenRequired();
-            final long minExistingGen = readers.isEmpty() ? current.getGeneration() : readers.get(0).getGeneration();
-            assert minReferencedGen >= minExistingGen :
+            assert minReferencedGen >= getMinFileGeneration() :
                 "deletion policy requires a minReferenceGen of [" + minReferencedGen + "] but the lowest gen available is ["
-                    + minExistingGen + "]";
-            final List<TranslogReader> unreferenced =
-                    readers.stream().filter(r -> r.getGeneration() < minReferencedGen).collect(Collectors.toList());
-            for (final TranslogReader unreferencedReader : unreferenced) {
+                    + getMinFileGeneration() + "]";
+            assert minReferencedGen <= currentFileGeneration() :
+                "deletion policy requires a minReferenceGen of [" + minReferencedGen + "] which is higher than the current generation ["
+                    + currentFileGeneration() + "]";
+
+            while (readers.isEmpty() == false && readers.get(0).getGeneration() < minReferencedGen) {
```

Comment:
can we use an iterator here instead?  it would be more clear to me if we'd do that..